### DISCUSSION TITLE
有没有方法配置代理不使用local dns

### DISCUSSION THREAD

**Discussion by mokitoo** [shadowsocks/shadowsocks-android#2727] [General]
**Is your feature request related to a problem? Please describe.**
目前的这个版本，vpn模式下，后台拿到的都是ip，看起来好像是local dns已经缓存了
看以前老版本里面，即使配置了远程dns，后台还是可以请求到域名

**Comment by mokitoo** [comment]
看了下好像是本地有个localDns的server，不知道这个localDns修改哪里可以关掉

**Comment by mokitoo** [comment]
试了下去掉local DNS proxy （把dnsRelay设置为false），好像直接就不能用了 @madeye  @Mygod 

ProxyInstance.kt
```kotlin
// local DNS proxy
            if (dnsRelay) try {
                URI("dns://${profile.remoteDns}")
            } catch (e: URISyntaxException) {
                throw BaseService.ExpectedExceptionWrapper(e)
            }.let { dns ->
                put(JSONObject().apply {
                    put("local_address", DataStore.listenAddress)
                    put("local_port", DataStore.portLocalDns)
                    put("local_dns_address", "local_dns_path")
                    put("remote_dns_address", dns.host ?: "0.0.0.0")
                    put("remote_dns_port", if (dns.port < 0) 53 else dns.port)
                    put("protocol", "dns")
                })
            }
```

**Reply by madeye** [reply]
It's expected. Local DNS proxy is required.

**Reply by mokitoo** [reply]
> It's expected. Local DNS proxy is required.

按照目前的原理，好像是local dns把域名发给远端请求dns.google , 返回的ip被local dns缓存，下一次就不需要重复请求远端dns了，不知道这么理解原理对不对。

如果是这样的话，那么远端的代理始终看不到代理的域名，这个有没有什么好的解决方案呢？

**Reply by madeye** [reply]
Set the route to `ALL`, then all the DNS queries should be sent to the remote DNS server.

**Reply by mokitoo** [reply]
> Set the route to `ALL`, then all the DNS queries should be sent to the remote DNS server.

全局代理会消耗很多不必要的流量，所以一般用的是bypass-lan-china  .....

**Reply by mokitoo** [reply]
> It's expected. Local DNS proxy is required.

看了下rust那边，Local DNS is optional , 不太明白为什么在android这里是必选项啊？

**Comment by mokitoo** [comment]
@Mygod  能帮忙解释一下，为什么dnsRelay不能设置为false嘛？ 我看udpFallback是可以设置为false的，但是data.proxy 设置为false，就没法用了

BaseService.kt
```kotlin
data.proxy!!.start(this,
                    File(Core.deviceStorage.noBackupFilesDir, "stat_main"),
                    File(configRoot, CONFIG_FILE),
                    if (udpFallback == null && data.proxy?.plugin == null) "tcp_and_udp" else "tcp_only")
            if (udpFallback?.plugin != null) throw ExpectedExceptionWrapper(IllegalStateException(
                    "UDP fallback cannot have plugins"))
            udpFallback?.start(this,
                    File(Core.deviceStorage.noBackupFilesDir, "stat_udp"),
                    File(configRoot, CONFIG_FILE_UDP),
                    "udp_only", false)
```

**Comment by Mygod** [comment]
The old behavior is harmful and removed. https://github.com/shadowsocks/shadowsocks-libev/issues/2627

**Reply by mokitoo** [reply]
> The old behavior is harmful and removed. [shadowsocks/shadowsocks-libev#2627](https://github.com/shadowsocks/shadowsocks-libev/issues/2627)

**Reply by mokitoo** [reply]
> The old behavior is harmful and removed. [shadowsocks/shadowsocks-libev#2627](https://github.com/shadowsocks/shadowsocks-libev/issues/2627)

@Mygod  再次仔细看了下，觉得优缺点各有，这种最好是做成一个option可选项，因为能拿到域名还是很有作用的，对于一些分流的规则

**Reply by Mygod** [reply]
Server side ACL?

**Reply by mokitoo** [reply]
> Server side ACL?

有一些流媒体的解锁需要拿到域名才行，比如Netflix , 通过remote proxy 去连接解锁dns

**Reply by mokitoo** [reply]
> Server side ACL?

解锁的dns一般会做限制，只允许remote proxy才能连接。所以这块sni功能还是很有用的，建议做成一个可选项。 @Mygod  @madeye

**Reply by Mygod** [reply]
I think setting remote DNS correctly on shadowsocks-android should resolve your issue.

See also: https://github.com/shadowsocks/shadowsocks-org/issues/156

**Reply by mokitoo** [reply]
> I think setting remote DNS correctly on shadowsocks-android should resolve your issue.
> 
> See also: [shadowsocks/shadowsocks-org#156](https://github.com/shadowsocks/shadowsocks-org/issues/156)

一些针对域名的审计规则，也是需要通过拿到域名来操作的，不只是流媒体的解锁。
简而言之，如果从易用性角度来说，把sni功能设置成可选项会是一个更好的选择
